File: 이정훈/6주차/[1912] 연속합.cpp
/*
* Author    : Lee JeongHun
* Date      : 2024.11.09(Sat)
* Runtime   : 24 ms
* Memory    : 2412 KB
* Algorithm : DP
*/

#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

int main()
{
    // 입력 처리
    int n;
    cin >> n;

    vector<int> array;
    array.reserve(n);
    for(int i = 0; i < n; ++i)
    {
        int input;
        cin >> input;

        array.push_back(input);
    }

    int sum = array[0];
    int answer = sum;
    for(int i = 1; i < n; ++i)
    {
        sum = sum < 0 ? array[i] : sum + array[i];
        answer = max(answer, sum);
    }

    // 정답 출력
    cout << answer;

    return 0;
}

File: 강병호/6week/[1912] 연속합.py
'''
* Author    : BYEONGHO KANG
* Date      : 2024.11.6(Thu)
* Runtime   : 84 ms
* Memory    : 39120 KB
* Algorithm : Dynamic Programming
'''

import sys
input = sys.stdin.read

data = input().splitlines()
n = int(data[0])
arr = list(map(int, data[1].split()))

dp = [0] * n
dp[0] = arr[0]

for i in range(1, n):
    dp[i] = max(arr[i], dp[i - 1] + arr[i])

print(max(dp))